Compute the commit order (inserts/deletes) on the entity level

mpdude opened the pull request:
Fixes some long-standing issues that have to do with commit order computation and cyclic associations.

When performing inserts or deletes of entities, the ORM tries to find a sequence of operations so that all foreign key constraints can be satisfied. Basically, before inserting an entity, all other entities it refers to have to be inserted. And before deleting an entity, all other entities that refer to it have to be deleted, or database-level `SET NULL` has to be used.

The current implementation for this works at the _entity class_ level. It checks which entity class refers to which other class(es), and based on this information, computes an ordering of entity classes. Then, all entities of the first class will be processed, then those of the second class and so on. The same ordering is used for deletions.

In the PRs #10531 and #10532, two examples are given why this is not sufficient: There may be references between entities _of the same class_, so that we can only insert those in the database by looking at their associations _at the entity level_. Or, we may have references between entities of two different classes that require us to go back and forth – insert one entity of the first class, before we can insert one for the second class, before we can go back to the first class for a third entity.

Then, there is a shortcoming in the current `CommitOrderCalculator`: It does not detect cycles at all. This refers to situations where we cannot find a commit order because _every_ single entity to be processed requires another one to come first. Cycles can be broken at NULLable foreign keys, where a NULL value can temporarily be used and the value updated at a later stage. This is called "extra updates". Places where such extra updates can be used to save the day can be found by backtracking in the commit order computation, in case cycles are found.

Last, the commit order for INSERTs and DELETEs needs to be based on different criteria: INSERTs can make use of extra updates and require NULLable foreign key definitions for this to work. DELETEs, however, currently do not make use of such extra updates. For them, whether the database can use "SET NULL" on foreign key deletions is the important factor.

To make it possible to review the changes, they were broken down into a series of independent PRs against a temporary branch. This PR here does a final merge of the result into `2.16.x`.

The individual PRs are:

* [x] #10592 
* [x] #10651
* [x] #10689
* [x] #10566
* [x] #10732
* [x] #10749
* [x] #10750
* [x] #10531
* [x] #10532
* [x] #10751 
* [x] #10755

(Potentially) fixed issues/closed PRs:
* [X] fixes #5538, although no complete test was provided 
* [X] fixes #6499, tests were picked from #7180
* [X] closes #6533, tests were picked from #7180
* [X] closes #8703, it was using the tests from #6499 as well
* [X] fixes #7866, using the same tests as #8703
* [X] fixes #7006, tests added via #10750
* [X] closes #7180, picked tests from their PR and extra tests from discussion comments via #10751
* [X] fixes #9192, added a test via #10755
* [X] closes #10348, adding tests via #10732
* [X] closes #8349, their tests being included in #10592

_Not_ fixed issues/PRs:
* #5109 – this is about extra unique constraints and/or performing DELETE before INSERT, beyond what the ORM requires itself to make foreign keys work; similar in #5742, #6776
* #5665, #10548 – those are about scheduling extra updates before DELETE, to null out foreign keys and not require a database-level `SET NULL`

(Probably) related issues/PR with unclear state and/or missing reproducers:

* #9376
* #9377

Definitely too big for this PR:
* Performance: Persisters no longer being able to batch inserts
* Performance: Optimizing the DFS to find a minimal feedback arc set/minimize the number of scheduled extra updates

mpdude committed:
Add test: Commit order calculation must happen on the entity level

Add tests for entity insertion and deletion that require the commit order calculation to happen on the entity level. This demonstrates the necessity for the changes in #10547.

This PR contains two tests with carefully constructed entity relationships, where we have a non-nullable `parent` foreign key relationships between entities stored in the same table.

Class diagram:

```mermaid
classDiagram
    direction LR
    class A
    class B
    A --> B : parent
    B --|> A
```

Object graph:

```mermaid
graph LR;
    b1 --> b2;
    b2 --> a;
    b3 --> b2;
```

 #### Situation before #10547

The commit order is computed by looking at the associations at the _table_ (= _class_) level. Once the ordering of _tables_ has been found, entities being mapped to the same table will be processed in the order they were given to `persist()` or `remove()`.

That means only a particular ordering of `persist()` or `remove()` calls (see comment in the test) works:

For inserts, the order must be `$a, $b2, $b1, $b3` (or `... $b3, $b1`), for deletions `$b1, $b3, $b2, $a`.

 #### Situation with entity-level commit order computation (as in #10547)

The ORM computes the commit order by considering associations at the _entity_ level. It will be able to find a working order by itself.

mpdude committed:
Add test: Entity insertions must not happen table-wise

Add tests for entity insertion and deletion that require writes to different tables in an interleaved fashion, and that have to re-visit a particular table.

 #### Background

In #10531, I've given an example where it is necessary to compute the commit order on the entity (instead of table) level.

Taking a closer look at the UoW to see how this could be achieved, I noticed that the current, table-level commit order manifests itself also in the API between the UoW and `EntityPersister`s.

 #### Current situation

The UoW computes the commit order on the table level. All entity insertions for a particular table are passed through `EntityPersister::addInsert()` and finally written through `EntityPersister::executeInserts()`.

 #### Suggested change

The test in this PR contains a carefully constructed set of four entities. Two of them are of the same class (are written to the same table), but require other entities to be processed first.

In order to be able to insert this set of entities, the ORM must be able to perform inserts for a given table repeatedly, interleaved with writing other entities to their respective tables.

rvanlaak commented:
I am a bit rusty on the topic regarding the debugging of the commit order, but based on looking at the `CommitOrderCalculator` your suggested changes look extremely promising in solving the long standing commit order bug.

derrabus commented on lib/Doctrine/ORM/Internal/CommitOrderCalculator.php:
```diff
@@ -118,47 +98,45 @@ public function sort()
     {
+        if ($vertex->state === VertexState::IN_PROGRESS) {
+            // This node is already on the current DFS stack. We've found a cycle!
+            throw new CycleDetectedException();
```
An exception should have a message.

mpdude replied:
Yes, will need to think about this a bit more. Probably the message should convey the exact cycle, ie which chain of entity associations is causing a foreign key loop

mpdude replied:
Fixed

derrabus commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2571,9 +2547,16 @@ public function lock($entity, int $lockMode, $lockVersion = null): void
     /**
      * Gets the CommitOrderCalculator used by the UnitOfWork to order commits.
      *
+     * @deprecated use createCommitOrderCalculator() instead
```
What's the reason for this deprecation? We usually don't introduce deprecations in bugfix releases and this one does not seem to be necessary.

mpdude replied:
Reverted

derrabus reviewed (changes requested):
I think I'll need to review this with more time than I have right now. Awesome work. 🤯

mpdude commented:
@derrabus Yes, this PR is much bigger than what I'd like to see to make reviews easy.

I did not find a good way to break it down into chunks that would make sense on their own and could be merged. Computing the order on the entity level with a broken CommitOrderCalculator does not work. Fixing the COC will break use cases that currently work as long as we don't consider individual entities.

What I could offer is that I try to break this down into pieces and guide you and everyone interested through reviews in dedicated PRs, with smaller scope each.

We'd need to merge all those PRs into a dedicated branch when they are approved. Tests for individual features probably can be working, but the full test suite won't pass until we're done.

In the end, the temporary branch could be merged and should roughly be equivalent to this PR here.

WDYT – would that help, or do you have other suggestions?

derrabus commented:
I could also offer to setup a Google Meet call where you can guide me and maybe @greg0ire and @beberlei through the changes. That might be more effective.

jhony1104 reviewed (dismissed):
This is no complete review, but since @mpdude [asked](https://github.com/doctrine/orm/pull/8349#issuecomment-1446836044) if I could look at it:

Only looking at the CommitOrderCalculator, unlike #8349, #7866, #8067, #7180, #9377 and others this PR correctly deals with circles using backtracing and if needed ignores nullable relations. Unlike some of the listet PRs it does not only work reliably for (simple) testcases, but I have also testet it in our product with ~130 tables (where for example #8349 didn't work).

Merging this would allow us to start using unmodified doctrine.

mpdude commented:
Rebased onto 2.15.x – given the nature of these changes, due to potential collateral damage this is nothing we should roll out in a bugfix release.

My plan is to break this down into a series of individual PRs to make reviews easier, as outlined in https://github.com/doctrine/orm/pull/10547#issuecomment-1449702387.

So, could somebody from the @doctrine group please create a dedicated branch (say, `entity-level-commit-order`?) off `2.15.x` for me? 

That way, all the individual PRs can happen here in doctrine/orm, but we collect the results on that branch until and can eventually merge the branch with results as a whole.

vctls commented on lib/Doctrine/ORM/Internal/TopologicalSort.php:
```diff
@@ -0,0 +1,146 @@
+ * TopologicalSort implements topological sorting, which is an ordering
+ * algorithm for directed graphs (DG) using a depth-first searching (DFS)
+ * to traverse the graph built in memory.
+ * This algorithm have a linear running time based on nodes (V) and edges
```
```suggestion
 * This algorithm has a linear running time based on nodes (V) and edges
```

greg0ire commented:
> So, could somebody from the https://github.com/doctrine group please create a dedicated branch (say, entity-level-commit-order?) off 2.15.x for me?

Done :)

mpdude commented:
Here we go: #10592 adds the new topological sort impl. Looking forward to reviews!

mpdude commented:
@greg0ire could you update the CI config so that it will run on the new branch as well?

greg0ire commented:
Done, I think.

mpdude committed:
Run workflows also for the `entity-level-commit-order` branch

greg0ire committed:
Merge pull request #10593 from mpdude/workflows

Run workflows also for the `entity-level-commit-order` branch

mpdude committed:
Prepare entity-level commit order computation in the UnitOfWork

This is the second chunk to break #10547 into smaller PRs suitable for reviewing. It prepares the `UnitOfWork` to work with a commit order computed on the entity level, as opposed to a class-based ordering as before.

#### Background

#10531 and #10532 show that it is not always possible to run `UnitOfWork::commit()` with a commit order given in terms of  entity _classes_. Instead it is necessary to process entities in an order computed on the _object_ level. That may include

* a particular order for multiple entities of the _same_ class
* a particular order for entities of _different_ classes, possibly even going back and forth (entity `$a1` of class `A`, then `$b` of class `B`, then `$a2` of class `A` – revisiting that class).

This PR is about preparing the `UnitOfWork` so that its methods will be able to perform inserts and deletions on that level of granularity. Subsequent PRs will deal with implementing that particular order computation.

#### Suggested change

Change the private `executeInserts` and `executeDeletions` methods so that they do not take a `ClassMetadata` identifying the class of entities that shall be processed, but pass them the list of entities directly.

The lists of entities are built in two dedicated methods. That happens basically as previously, by iterating over `$this->entityInsertions` or `$this->entityDeletions` and filtering those by class.

#### Potential (BC breaking?) changes that need review scrutiny

* `\Doctrine\ORM\Persisters\Entity\EntityPersister::addInsert()` was previously called multiple times, before the insert would be performed by a call to `\Doctrine\ORM\Persisters\Entity\EntityPersister::executeInserts()`. With the changes here, this batching effectively no longer takes place. `executeInserts()` will always see one entity/insert only, and there may be multiple `executeInserts()` calls during a single `UoW::commit()` phase.
* The caching in `\Doctrine\ORM\Cache\Persister\Entity\AbstractEntityPersister` previously would cache entities from the last executed insert batch only. Now it will collect entities across multiple batches. I don't know if that poses a problem.
* Overhead in `\Doctrine\ORM\Persisters\Entity\BasicEntityPersister::executeInserts` is incurred multiple times; that may, however, only be about SQL statement preparation and might be salvageable.
* The `postPersist` event previously was not dispatched before _all_ instances of a particular entity class had been written to the database. Now, it will be dispatched immediately after every single entity that has been inserted.

mpdude committed:
More precisely state conditions for the postPersist event

greg0ire committed:
Merge pull request #10651 from mpdude/unit-of-work-schedule-single-entities

Prepare entity-level commit order computation in the `UnitOfWork`

mpdude committed:
Add a new topological sort implementation (#10592)

This is the first chunk to break #10547 into smaller PRs suitable for reviewing. It adds a new topological sort implementation.

 #### Background

Topological sort is an algorithm that sorts the vertices of a directed acyclic graph (DAG) in a linear order such that for every directed edge from vertex A to vertex B, vertex A comes before vertex B in the ordering. This ordering is called a topological order.

Ultimately (beyond the scope of this PR), in the ORM we'll need this to find an order in which we can insert new entities into the database. When one entity needs to refer to another one by means of a foreign key, the referred-to entity must be inserted before the referring entity. Deleting entities is similar.

A topological sorting can be obtained by running a depth first search (DFS) on the graph. The order in which the DFS finishes on the vertices is a topological order. The DFS is possible iif there are no cycles in the graph. When there are cycles, the DFS will find them.

For more information about topological sorting, as well as a description of an DFS-based topological sorting algorithm, see https://en.wikipedia.org/wiki/Topological_sorting.

 #### Current situation

There is a DFS-based topological sorting implemented in the `CommitOrderCalculator`. This implementation has two kinks:

1. It does not detect cycles

When there is a cycle in the DAG that cannot be resolved, we need to know about it. Ultimately, this means we will not be able to insert entities into the database in any order that allows all foreign keys constraints to be satisfied.

If you look at `CommitOrderCalculator`, you'll see that there is no code dealing with this situation.

2. It has an obscure concept of edge "weights"

To me, it is not clear how those are supposed to work. The weights are related to whether a foreign key is nullable or not, but can (could) be arbitrary integers. An edge will be ignored if it has a higher (lower) weight than another, already processed edge... 🤷🏻?

 #### Suggested change

In fact, when inserting entities into the database, we have two kinds of foreign key relationships: Those that are `nullable`, and those that are not.

Non-nullable foreign keys are hard requirements: Referred-to entities must be inserted first, no matter what. These are "non-optional" edges in the dependency graph.

Nullable foreign keys can be set to `NULL` when first inserting an entity, and then coming back and updating the foreign key value after the referred-to (related) entity has been inserted into the database. This is already implemented in `\Doctrine\ORM\UnitOfWork::scheduleExtraUpdate`, at the expense of performing one extra `UPDATE` query after all the `INSERT`s have been processed. These edges are "optional".

When finding a cycle that consists of non-optional edges only, treat it as a failure. We won't be able to insert entities with a circular dependency when all foreign keys are non-NULLable.

When a cycle contains at least one optional edge, we can use it to break the cycle: Use backtracking to go back to the point before traversing the last _optional_ edge. This omits the edge from the topological sort order, but will cost one extra UPDATE later on.

To make the transition easier, the new implementation is provided as a separate class, which is marked as `@internal`.

 #### Outlook

Backtracking to the last optional edge is the simplest solution for now. In general, it might be better to find _another_ (optional) edge that would also break the cycle, if that edge is also part of _other_ cycles.

Remember, every optional edge skipped costs an extra UPDATE query later on. The underlying problem is known as the "minimum feedback arc set" problem, and is not easily/efficiently solvable. Thus, for the time being, picking the nearest optional edge seems to be reasonable.

mpdude committed:
Compute entity-level commit order for entity insertions

This is the third step to break https://github.com/doctrine/orm/pull/10547 into smaller PRs suitable for reviewing. It uses the new topological sort implementation from #10592 and the refactoring from #10651 to compute the UoW's commit order for entity insertions not on the entity class level, but for single entities and their actual dependencies instead.

 #### Current situation

`UnitOfWork::getCommitOrder()` would compute the entity sequence on the class level with the following code:

https://github.com/doctrine/orm/blob/70477d81e96c0044ad6fd8c13c37b2270d082792/lib/Doctrine/ORM/UnitOfWork.php#L1310-L1325

 #### Suggested change

* Instead of considering the classes of all entities that need to be inserted, updated or deleted, consider the new (inserted) entities only. We only need to find a sequence in situations where there are foreign key relationships between two _new_ entities.
* In the dependency graph, add edges for all to-one association target entities.
* Make edges "optional" when the association is nullable.

 #### Test changes

I have not tried to fully understand the few changes necessary to fix the tests. My guess is that those are edge cases where the insert order changed and we need to consider this during clean-up.

Keep in mind that many of the functional tests we have assume that entities have IDs assigned in the order that they were added to the EntityManager. That does not change – so the order of entities is generally stable, equal to the previous implementation.

mpdude committed:
Commit order for removals has to consider `SET NULL`, not `nullable` (#10566)

When computing the commit order for entity removals, we have to look out for `@ORM\JoinColumn(onDelete="SET NULL")` to find places where cyclic associations can be broken.

 #### Background

The UoW computes a "commit order" to find the sequence in which tables shall be processed when inserting entities into the database or performing delete operations.

For the insert case, the ORM is able to schedule _extra updates_ that will be performed after all entities have been inserted. Associations which are configured as `@ORM\JoinColumn(nullable=true, ...)` can be left as `NULL` in the database when performing the initial `INSERT` statements, and will be updated once all new entities have been written to the database. This can be used to break cyclic associations between entity instances.

For removals, the ORM does not currently implement up-front `UPDATE` statements to `NULL` out associations before `DELETE` statements are executed. That means when associations form a cycle, users have to configure `@ORM\JoinColumn(onDelete="SET NULL", ...)` on one of the associations involved. This transfers responsibility to the DBMS to break the cycle at that place.

_But_, we still have to perform the delete statements in an order that makes this happen early enough. This may be a _different_ order than the one required for the insert case. We can find it _only_ by looking at the `onDelete` behaviour. We must ignore the `nullable` property, which is irrelevant, since we do not even try to `NULL` anything.

 #### Example

Assume three entity classes `A`, `B`, `C`. There are unidirectional one-to-one associations `A -> B`, `B -> C`, `C -> A`. All those associations are `nullable= true`.

Three entities `$a`, `$b`, `$c` are created from these respective classes and associations are set up.

All operations `cascade` at the ORM level. So we can test what happens when we start the operations at the three individual entities, but in the end, they will always involve all three of them.

_Any_ insert order will work, so the improvements necessary to solve #10531 or #10532 are not needed here. Since all associations are between different tables, the current table-level computation is good enough.

For the removal case, only the `A -> B` association has `onDelete="SET NULL"`. So, the only possible execution order is `$b`, `$c`, `$a`. We have to find that regardless of where we start the cascade operation.

The DBMS will set the `A -> B` association on `$a` to `NULL` when we remove `$b`. We can then remove `$c` since it is no longer being referred to, then `$a`.

 #### Related cases

These cases ask for the ORM to perform the extra update before the delete by itself, without DBMS-level support:
* #5665
* #10548

mpdude committed:
Remove the now unused UnitOfWork::getCommitOrder() method

mpdude committed:
Add a test case to show #10348 has been fixed by #10566

This is part of the series of issues fixed by #10547. In particular, the changes from #10566 were relevant.

See #10348 for the bug description.

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

greg0ire committed:
Merge pull request #10732 from mpdude/10348-fixed

Add a test case to show #10348 has been fixed

greg0ire committed:
Merge pull request #10689 from mpdude/insert-entity-level-topo-sort

mpdude committed:
Add tests to show #6499 has been fixed

@frikkle was the first to propose these tests in #6533.
@rvanlaak followed up in #8703, making some adjustments.

Co-authored-by: Gabe van der Weijde <gabe.vanderweijde@triasinformatica.nl>
Co-authored-by: Richard van Laak <rvanlaak@gmail.com>
Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

mpdude committed:
Add test to show #7006 has been fixed

mpdude committed:
Update tests/Doctrine/Tests/ORM/Functional/Ticket/GH7006Test.php

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

greg0ire committed:
Merge pull request #10750 from mpdude/7006-fixed

Add test to show #7006 has been fixed

mpdude committed:
Add test to show #7180 has been fixed

Tests suggested in https://github.com/doctrine/orm/pull/7180#issuecomment-380841413 and https://github.com/doctrine/orm/pull/7180#issuecomment-381067448 by @arnaud-lb.

Co-authored-by: Arnaud Le Blanc <arnaud.lb@gmail.com>

greg0ire committed:
Merge pull request #10749 from mpdude/6499-fixed

Add tests to show #6499 has been fixed

mpdude committed:
Add test to show #9192 has been fixed

This test implements the situation described in #9192. The commit order computation will be fixed by #10547.

greg0ire committed:
Merge pull request #10755 from mpdude/9192-fixed

Add test to show #9192 has been fixed

greg0ire committed:
Merge pull request #10751 from mpdude/7180-fixed

Add test to show #7180 has been fixed

greg0ire committed:
Merge remote-tracking branch 'origin/2.15.x' into entity-level-commit-order

greg0ire committed:
Merge pull request #10792 from greg0ire/entity-level-commit-order

Entity level commit order

greg0ire committed:
Merge pull request #10531 from mpdude/commit-order-must-be-entity-level

Add test: Commit order calculation must happen on the entity level

greg0ire committed:
Merge pull request #10532 from mpdude/entity-persister-must-not-batch

greg0ire reviewed (commented):
There are conflicts

mpdude committed:
Merge remote-tracking branch 'origin/2.16.x' into commit-order-entity-level

mpdude commented:
Force-pushed this branch to be the same as `entity-level-commit-order` where all the intermediate work took place. Also, merged 2.16.x and resolved conflicts 🤞🏻

greg0ire commented:
![y u no use git hooks](https://i.imgflip.com/7qulwo.jpg)

mpdude committed:
Avoid unnecessarily passing entity lists into executeDeletions/executeInserts

greg0ire commented on .github/workflows/coding-standards.yml:
```diff
@@ -4,6 +4,7 @@ on:
   pull_request:
     branches:
       - "*.x"
+      - entity-level-commit-order
```
Isn't the plan to merge this PR and then remove that branch?

mpdude replied:
Yes. I had those things in `.github/workflows` in mind, but thought we cannot safely remove them unless this PR is merged?

greg0ire replied:
Yes exactly… so now you should push a new commit to remove those, and the PR should be ready to merge, right?

mpdude replied:
Did that. Surprised to see that tests etc. still being run for that new commit.

mpdude replied:
D'oh. I get it. They run for the PR branch, no longer need to run for the other one.

mpdude committed:
Remove references to the temporary branch in workflow definitions

mpdude commented:
@derrabus Maybe you need to approve before we can merge this :-)

derrabus commented on lib/Doctrine/ORM/Internal/TopologicalSort.php:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * Builds up the result during the DFS.
+     *
+     * @psalm-var list<object>
```
I think, we can un-prefix simple generics like `list<>` and `array<>` nowadays.

```suggestion
     * @var list<object>
```

mpdude replied:
Updated, and found one more case

derrabus commented on lib/Doctrine/ORM/Internal/TopologicalSort.php:
```diff
@@ -0,0 +1,165 @@
+     *
+     * @psalm-return list<object>
+     */
+    public function sort()
```
```suggestion
    public function sort(): array
```

mpdude replied:
done

derrabus reviewed (dismissed):
LGTM

mpdude committed:
Un-prefix simple generics like `list<>` and `array<>`

... as suggested in GH review.

mpdude commented:
My push dismissed your reviews.

mpdude committed:
Deprecate classes related to old commit order computation

mpdude commented:
Disclosure: I've added a commit bb21865cba0475d74ada1f487e4c2fef59256ee4 that did not go through the review process on the dedicated `entity-level-commit-order` branch.

greg0ire commented:
The extra commit looks correct to me.

derrabus commented on lib/Doctrine/ORM/Internal/CommitOrder/Edge.php:
```diff
@@ -27,6 +29,13 @@ final class Edge
 
     public function __construct(string $from, string $to, int $weight)
     {
+        Deprecation::triggerIfCalledFromOutside(
+            'doctrine/orm',
+            'https://github.com/doctrine/orm/pull/10547',
+            'The %s class is deprecated and will be removed in ORM 3.0',
+            self::class
+        );
```
If the class is deprecated, it should be annotated with `@deprecated`. That being said, it's already `@internal`, so we might as well remove it right away?

mpdude replied:
I'd love to see it being removed right away, but the `CommitOrderCalculator` depends on it and was itself not marked as `@internal` (it only lived in a `...\Internal\...` namespace).

So, I'll keep the classes for now but added `@deprecated` annotations.

mpdude committed:
Add `@deprecated` annotations in addition to runtime deprecation notices

sylfabre commented:
Hello @mpdude 

Thank you for addressing this long-standing issue 👍 

I tried your branch as per your comments on https://github.com/doctrine/orm/pull/10714 & https://github.com/doctrine/orm/pull/10716
Unfortunately, I get an exception in `Statement.php` with this message: 
```
 [PDOException (42000)] SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '?, ?, ?, ?, ?)' at line 1
```

This comes from the insert statement `INSERT INTO payment_settings (updated_at, retribution_method, id, created_at, organization_id) VALUES (?, ?, ?, ?, ?)` which is done with an empty params array 😬 

The same code works using `doctrine/orm` v2.15.3

So I cannot confirm if your PR solves my issues https://github.com/doctrine/orm/issues/10713 & https://github.com/doctrine/orm/issues/10715

Let me know how I may help you.
Our database contains over 250 tables so it can be a strong real-world test!

mpdude commented:
@sylfabre Can you provide a test (as a PR against the `entity-level-commit-order` branch) that demonstrates the issue? 

The test outlined in #10713 works directly with the `CommitOrderCalculator`, a class replaced by this PR here.

sylfabre commented:
Hello @mpdude 

I'm sorry my last message was a false positive 👌 

After checking my codebase, I discovered a postPersist listener triggering a flush operation. It leads to trying to insert two times the same entity: the first time works, the second fails because the entity changeset is now empty so the insert query is missing a lot of params. I guess flushing within a postPersist listener is a bad practice and we are just lucky it works today!

Thank you again for your work

mpdude committed:
Fix commit order computation for self-referencing entities with application-provided IDs

This excludes such associations from the commit order computation, since the foreign key constraint will be satisfied when inserting the row.

See https://github.com/doctrine/orm/pull/10735/ for more details about this edge case.

mpdude commented:
Added f01d107edc5bebbd3d628a7c6cb1934339aea435 to address a new edge case supported in 2.16 since #10735.

mpdude committed:
Exclude deprecated classes from Psalm checks (until 3.0)

greg0ire commented:
Congrats on this huge piece of work, and welcome back!

rvanlaak commented:
Congrats! This is an awesome step in the reliability of the commit order.

mpdude commented:
Thanks to everyone involved and supporting this over the last months with many reviews in a very constructive and responsive manner – especially @greg0ire, @SenseException and @derrabus!

polaeskandar commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -437,32 +434,37 @@ public function commit($entity = null)
-                    $this->executeUpdates($class);
-                }
+                // Updates do not need to follow a particular order
+                $this->executeUpdates();
```
https://github.com/doctrine/orm/issues/10864
